docs/topics/gradle/gradle-compiler-options.md

```diff
@@ -15,8 +15,52 @@ in the [Working with command-line compiler](command-line.md) tutorial.
 
 Kotlin compilers have a number of options for tailoring the compiling process.
 
-Using a build script, you can specify additional compilation options. Use the `compilerOptions` property of a Kotlin compilation task for it. 
-For example:
+When you use the Kotlin Gradle plugin in your project, the Gradle DSL allows comprehensive configuration of compiler options. 
+The Gradle DSL is available for [Kotlin Multiplatform](multiplatform-dsl-reference.md) and [JVM/Android](#target-the-jvm) projects.
+
+With the Gradle DSL, you can configure compiler options within the build script at three levels: 
+* **Extension level in top-level blocks:** for all the targets and shared source sets.
+* **Target level in target-level blocks:** for a specific target.
+* **Compilation unit level:** usually a specific compilation task.
+
+![Kotlin compiler options levels](compiler-options-levels.svg){width=700}
+
+The settings at a higher level are used as a convention (default) for a lower level:
+
+* The values of extension compiler options are the default for target compiler options, including shared source sets
+  like `commonMain`, `nativeMain`, and `commonTest`.
+* The values of target compiler options are used as the default for compilation unit (task) compiler options, for
+  example, `compileKotlinJvm` and `compileTestKotlinJvm` tasks.
+
+In turn, configurations made at a lower level override related settings at a higher level:
+
+* Task-level compiler options override related configurations at the target or the extension level.
+* Target-level compiler options override related configurations at the extension level.
+
+To confirm which level of compiler arguments is applied to the compilation, use the `DEBUG` level of Gradle [logging](https://docs.gradle.org/current/userguide/logging.html).
+Gradle is making these properties traceable, so understanding what overrides your values will soon be easier.
+
+> If you are a third-party plugin author, it is recommended that you apply your configuration on the project level to avoid 
+> overriding issues. Also, it is recommended that you document such configuration on your side explicitly.
+>
+{style="tip"}
+
+### Extension level
+
+You can configure compiler options for all the targets and source sets
+in the `compilerOptions {}` block at the top level:
+
+```kotlin
+kotlin {
+    compilerOptions {
+        optIn.add("kotlin.RequiresOptIn")
+    }
+}    
+```
+
+If you want to configure a plugin of a target different from JVM/Android and [Kotlin Multiplatform](multiplatform-dsl-reference.md), 
+use the `compilerOptions {}` property of the corresponding Kotlin compilation task. The following examples 
+show how to set this configuration up in both Kotlin and Groovy DSLs:
```

Comment:
It looks should not belong to "Extension level" section - maybe could be a part of "Compilation Unit" section